### PR TITLE
NIOPerformanceTester: Add DeadlineNowBenchmark for NIODeadline.now()

### PR DESCRIPTION
**Pull request by simonjbeaumont**
### Motivation:

`NIODeadline.now()` is a function that is called many times in typical workloads. It is currently implemented in terms of `DispatchTime.now()`, which is provided by libdispatch, on all platforms. The underlying implementation has a lot of branches to offer a more flexible API than we make use of. There might be a chance to improve the performance by dropping down to a lower-level implementation within NIO itself but we need a baseline measurement of performance to assess any future proposed improvements.

### Modifications:

Add performance benchmark for `NIODeadline.now()`.

### Result:

`NIOPerformanceTester` has benchmark test for `NIODeadline.now()`.

### CODE REVIEW

**Review by glbrntt — APPROVED** [review]
LGTM. I'm interested to see what perf gains can be made from looking into this!

**Comment by glbrntt** [comment]
@swift-nio-bot test perf please

**Comment by simonjbeaumont** [comment]
@glbrntt @Lukasa the CI is failing with:

```
13:20:25 .FileNotFoundError: [Errno 2] No such file or directory: './docker/docker-compose.2004.54.yaml'
```

I noticed that this was removed recently, in #2080 (although, maybe not deliberately since the commit message is only about dropping 5.2 and 5.3).

I guess we either need this Docker Compose file for 5.4 back, or we need to update the CI job to use a newer one?

**Comment by glbrntt** [comment]
```
13:20:24 + docker-compose -f docker/docker-compose.yaml -f docker/docker-compose.2004.54.yaml -p swift-nio2-performance-prb build --pull
13:20:25 .FileNotFoundError: [Errno 2] No such file or directory: './docker/docker-compose.2004.54.yaml'
13:20:25 Build step 'Execute shell' marked build as failure
```

CI is expecting `docker-compose.2004.54.yaml` but we recently moved Swift 5.4 CI to 18.04 (https://github.com/apple/swift-nio/pull/2080) so that file no longer exists which is why perf tests now fail.

I suspect we actually want to use Swift 5.6.

**Comment by simonjbeaumont** [comment]
@glbrntt I'll open a PR for that separately.

**Comment by simonjbeaumont** [comment]
> I suspect we actually want to use Swift 5.6.

Who can update the CI job to make use of 5.6 one?

In the meantime should we reinstate the 5.4 Bionic?

**Comment by Lukasa** [comment]
@yim-lee can you update the performance jobs to use 5.6 instead of 5.4? Thankyou!

File: Sources/NIOPerformanceTester/DeadlineNowBenchmark.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    func run() -> Int {
+        var counter: UInt64 = 0
+        for _ in 0..<self.iterations {
+            let now = NIODeadline.now().uptimeNanoseconds
+            counter &+= now
+
+        }
+        return 1
```
Comment (Lukasa) [inline Sources/NIOPerformanceTester/DeadlineNowBenchmark.swift:37]: Can I suggest we return `counter` here? That way the benchmarks will be less inclined to kill the code.

Comment (simonjbeaumont) [reply Sources/NIOPerformanceTester/DeadlineNowBenchmark.swift:37]: Sure, I think I saw another test returning `1` here. Do I understand correctly that this return value has no meaningful use, but is just used to ensure that code doesn't get optimised away?

Comment (simonjbeaumont) [reply Sources/NIOPerformanceTester/DeadlineNowBenchmark.swift]: Addressed in https://github.com/apple/swift-nio/pull/2117/commits/edffb7d202d004c86c900ebe1bc85c7930482857.

Comment (Lukasa) [reply Sources/NIOPerformanceTester/DeadlineNowBenchmark.swift]: @simonjbeaumont Yes: the result is thrown away, but the call site doesn't know that.

**Comment by simonjbeaumont** [comment]
@swift-nio-bot test perf please

**Comment by simonjbeaumont** [comment]
> significant differences found

These are a result of the move to Swift 5.6 as the image for the perf test presumably since nothing in this PR will affect those benchmarks.

**Review by Lukasa — APPROVED** [review]

**Pull request merged**